File: internal/eventstore/eventstore.go
package eventstore

import (
	"context"
	"fmt"
	"log"
	"sync"
	"time"

	"github.com/junkawasaki/actordb-dokigoto/pkg/config"
)

// Event represents a single event in the event store
// Merkle DAG: sha256:event_v1 - Single-writer actor event persistence
type Event struct {
	AggregateID   string                 `json:"aggregate_id"`
	Sequence      int64                  `json:"sequence"`
	EventType     string                 `json:"event_type"`
	Data          []byte                 `json:"data"`
	Timestamp     time.Time              `json:"timestamp"`
	EventTime     time.Time              `json:"event_time"` // For late event handling
	AggregateType string                 `json:"aggregate_type"`
	Metadata      map[string]interface{} `json:"metadata,omitempty"`
}

// ActorState represents the current state of an actor
type ActorState struct {
	AggregateID   string
	LastSequence  int64
	LastTimestamp time.Time
	SnapshotData  []byte
	SnapshotSeq   int64
}

// WriteResult contains the result of a write operation
type WriteResult struct {
	AggregateID string
	Sequence    int64
	Timestamp   time.Time
	Success     bool
	Error       error
}

// EventStore handles actor-based append-only event storage
// Process Network Node: write_aggregate
// Dependencies: []
// Outputs: [event_stream]
// SLO: p99_latency_100ms
type EventStore struct {
	config     config.EventStoreConfig
	actors     map[string]*ActorState
	actorsMu   sync.RWMutex
	eventLog   []Event // In-memory for MVP; would be persistent storage
	eventLogMu sync.RWMutex
	running    bool
	ctx        context.Context
	cancel     context.CancelFunc
}

// New creates a new EventStore instance
func New(cfg config.EventStoreConfig) (*EventStore, error) {
	ctx, cancel := context.WithCancel(context.Background())

	es := &EventStore{
		config:   cfg,
		actors:   make(map[string]*ActorState),
		eventLog: make([]Event, 0),
		running:  false,
		ctx:      ctx,
		cancel:   cancel,
	}

	return es, nil
}

// Start begins the EventStore operation
func (es *EventStore) Start(ctx context.Context) error {
	es.running = true
	log.Printf("EventStore started with data_dir: %s", es.config.DataDir)
	return nil
}

// Stop shuts down the EventStore
func (es *EventStore) Stop() {
	es.running = false
	es.cancel()
	log.Println("EventStore stopped")
}

// WriteEvent appends an event to an actor's event stream
// Ensures single-writer serialization per actor
func (es *EventStore) WriteEvent(ctx context.Context, event Event) (*WriteResult, error) {
	if !es.running {
		return nil, fmt.Errorf("eventstore not running")
	}

	es.actorsMu.Lock()
	actor, exists := es.actors[event.AggregateID]
	if !exists {
		actor = &ActorState{
			AggregateID:  event.AggregateID,
			LastSequence: 0,
		}
		es.actors[event.AggregateID] = actor
	}
	es.actorsMu.Unlock()

	// Validate sequence (single-writer check)
	expectedSeq := actor.LastSequence + 1
	if event.Sequence != expectedSeq {
		return &WriteResult{
			AggregateID: event.AggregateID,
			Success:     false,
			Error:       fmt.Errorf("sequence mismatch: expected %d, got %d", expectedSeq, event.Sequence),
		}, nil
	}

	// Set timestamp if not provided
	if event.Timestamp.IsZero() {
		event.Timestamp = time.Now()
	}

	// Append to event log
	es.eventLogMu.Lock()
	es.eventLog = append(es.eventLog, event)
	es.eventLogMu.Unlock()

	// Update actor state
	actor.LastSequence = event.Sequence
	actor.LastTimestamp = event.Timestamp

	// Check if snapshot is needed
	if event.Sequence%es.config.SnapshotInterval == 0 {
		if err := es.createSnapshot(event.AggregateID, event.Sequence); err != nil {
			log.Printf("Failed to create snapshot for %s: %v", event.AggregateID, err)
		}
	}

	result := &WriteResult{
		AggregateID: event.AggregateID,
		Sequence:    event.Sequence,
		Timestamp:   event.Timestamp,
		Success:     true,
	}

	return result, nil
}

// ReadEvents reads events for an actor from a given sequence
func (es *EventStore) ReadEvents(ctx context.Context, aggregateID string, fromSeq int64) ([]Event, error) {
	if !es.running {
		return nil, fmt.Errorf("eventstore not running")
	}

	es.eventLogMu.RLock()
	defer es.eventLogMu.RUnlock()

	var events []Event
	for _, event := range es.eventLog {
		if event.AggregateID == aggregateID && event.Sequence >= fromSeq {
			events = append(events, event)
		}
	}

	return events, nil
}

// GetActorState returns the current state of an actor
func (es *EventStore) GetActorState(aggregateID string) (*ActorState, error) {
	es.actorsMu.RLock()
	defer es.actorsMu.RUnlock()

	actor, exists := es.actors[aggregateID]
	if !exists {
		return nil, fmt.Errorf("actor %s not found", aggregateID)
	}

	// Return a copy to prevent external modification
	state := *actor
	return &state, nil
}

// createSnapshot creates a snapshot for an actor (MVP: in-memory)
func (es *EventStore) createSnapshot(aggregateID string, sequence int64) error {
	// MVP: Just log the snapshot creation
	// In production: serialize actor state and persist
	log.Printf("Creating snapshot for actor %s at sequence %d", aggregateID, sequence)
	return nil
}

// GetAllEvents returns all events (for testing/debugging)
func (es *EventStore) GetAllEvents() []Event {
	es.eventLogMu.RLock()
	defer es.eventLogMu.RUnlock()

	events := make([]Event, len(es.eventLog))
	copy(events, es.eventLog)
	return events
}

File: demo/nextjs-libsql/lib/actordb/eventstore.ts
// Merkle DAG: eventstore_libsql -> libsql_storage
// LibSQL-based event store for ActorDB demo
// Process Network Node: write_aggregate (LibSQL variant)

import { getDatabaseClient } from '../database/config'

export interface Event {
  aggregateId: string
  aggregateType: string
  sequence: number
  eventType: string
  data: any
  timestamp: Date
  eventTime: Date
  metadata?: Record<string, any>
}

export interface WriteResult {
  aggregateId: string
  sequence: number
  success: boolean
  timestamp: Date
  error?: string
}

export interface ActorState {
  aggregateId: string
  lastSequence: number
  lastTimestamp: Date
}

export class LibSQLEventStore {
  private db = getDatabaseClient()
  private eventListeners: Set<(event: Event) => void> = new Set()

  // WriteEvent appends an event to an actor's event stream
  async writeEvent(event: Event): Promise<WriteResult> {
    try {
      // Get the next sequence number
      const lastSeqResult = await this.db.execute({
        sql: 'SELECT MAX(sequence) as last_seq FROM events WHERE aggregate_id = ?',
        args: [event.aggregateId]
      })

      const lastSeq = lastSeqResult.rows[0]?.last_seq || 0
      const nextSeq = lastSeq + 1

      // Validate sequence number
      if (event.sequence !== nextSeq) {
        return {
          aggregateId: event.aggregateId,
          sequence: 0,
          success: false,
          timestamp: new Date(),
          error: `Invalid sequence number. Expected ${nextSeq}, got ${event.sequence}`
        }
      }

      // Insert the event
      await this.db.execute({
        sql: `INSERT INTO events (
          aggregate_id, aggregate_type, sequence, event_type, data,
          timestamp, event_time, metadata
        ) VALUES (?, ?, ?, ?, ?, ?, ?, ?)`,
        args: [
          event.aggregateId,
          event.aggregateType,
          event.sequence,
          event.eventType,
          JSON.stringify(event.data),
          event.timestamp.toISOString(),
          event.eventTime.toISOString(),
          event.metadata ? JSON.stringify(event.metadata) : null
        ]
      })

      // Broadcast the event
      this.broadcastEvent(event)

      return {
        aggregateId: event.aggregateId,
        sequence: event.sequence,
        success: true,
        timestamp: event.timestamp
      }

    } catch (error) {
      return {
        aggregateId: event.aggregateId,
        sequence: 0,
        success: false,
        timestamp: new Date(),
        error: error instanceof Error ? error.message : 'Unknown error'
      }
    }
  }

  // ReadEvents reads events for an actor from a given sequence
  async readEvents(aggregateId: string, fromSeq: number = 1): Promise<Event[]> {
    const result = await this.db.execute({
      sql: `SELECT * FROM events
            WHERE aggregate_id = ? AND sequence >= ?
            ORDER BY sequence ASC`,
      args: [aggregateId, fromSeq]
    })

    return result.rows.map(row => ({
      aggregateId: row.aggregate_id as string,
      aggregateType: row.aggregate_type as string,
      sequence: row.sequence as number,
      eventType: row.event_type as string,
      data: JSON.parse(row.data as string),
      timestamp: new Date(row.timestamp as string),
      eventTime: new Date(row.event_time as string),
      metadata: row.metadata ? JSON.parse(row.metadata as string) : undefined
    }))
  }

  // GetActorState returns the current state of an actor
  async getActorState(aggregateId: string): Promise<ActorState | null> {
    const result = await this.db.execute({
      sql: `SELECT MAX(sequence) as last_seq, MAX(timestamp) as last_time
            FROM events WHERE aggregate_id = ?`,
      args: [aggregateId]
    })

    const row = result.rows[0]
    if (!row || !row.last_seq) {
      return null
    }

    return {
      aggregateId,
      lastSequence: row.last_seq as number,
      lastTimestamp: new Date(row.last_time as string)
    }
  }

  // GetAllEvents returns all events (for debugging/testing)
  async getAllEvents(limit: number = 1000): Promise<Event[]> {
    const result = await this.db.execute({
      sql: `SELECT * FROM events ORDER BY id DESC LIMIT ?`,
      args: [limit]
    })

    return result.rows.map(row => ({
      aggregateId: row.aggregate_id as string,
      aggregateType: row.aggregate_type as string,
      sequence: row.sequence as number,
      eventType: row.event_type as string,
      data: JSON.parse(row.data as string),
      timestamp: new Date(row.timestamp as string),
      eventTime: new Date(row.event_time as string),
      metadata: row.metadata ? JSON.parse(row.metadata as string) : undefined
    })).reverse() // Return in chronological order
  }

  // Subscribe adds a new event listener
  subscribe(listener: (event: Event) => void): () => void {
    this.eventListeners.add(listener)

    // Return unsubscribe function
    return () => {
      this.eventListeners.delete(listener)
    }
  }

  private broadcastEvent(event: Event): void {
    for (const listener of this.eventListeners) {
      try {
        listener(event)
      } catch (err) {
        console.error(`Error in event listener: ${err}`)
      }
    }
  }
}

// Singleton instance for the application
let eventStoreInstance: LibSQLEventStore | null = null

export function getEventStore(): LibSQLEventStore {
  if (!eventStoreInstance) {
    eventStoreInstance = new LibSQLEventStore()
  }
  return eventStoreInstance
}
